python/deltalake/table.py

```diff
@@ -313,6 +313,19 @@ def vacuum(
 
         return self._table.vacuum(dry_run, retention_hours, enforce_retention_duration)
 
+    def optimize(
+            self,
+            partition_filters: Optional[List[Tuple[str, str, Any]]] = None,
+            target_size: Optional[int] = None,
+    ) -> Dict[str, Any]:
+        """
+        :param partition_filters: the partition filters that will be used for getting the matched files
+        :param target_size: desired file size after bin-packing files
+        :return: the metrics from optimize
+        """
+        metrics = self._table.optimize(partition_filters, target_size)
+        return json.loads(metrics)
```

Comment:
It would be nice to have this be typed for the sake of autocompletion, but doing that in Rust right now involves a bit of boilerplate. One lightweight option is https://docs.python.org/3.8/library/typing.html#typing.TypedDict, but that's Python>=3.8 only. We still have a month or two until 3.7 is EOL.

Reply:
Perhaps something like:

```python
if TYPE_CHECKING and sys.version_info().minor >= 8:
    from typing import TypedDict

    class MetricDetails(TypeDict):
        min: int
        max: int
        avg: float
        total_files: int
        total_size: int
    
    class OptimizeMetrics(TypeDict):
        num_files_added: int
        num_files_removed: int
        files_added: MetricDetails
        files_removed: MetricDetails
        partitions_optimized: int
        num_batches: int
        total_considered_files: int
        total_files_skipped: int
        preserve_insertion_order: bool
```

Then you can modify the signature:


```python
def optimize(
            self,
            partition_filters: Optional[List[Tuple[str, str, Any]]] = None,
            target_size: Optional[int] = None,
    ) -> "OptimizeMetrics":
```

Reply:
(I'm happy to leave this for a follow up later FYI)

Reply:
ok. I'll leave this in current PR.